mlir/lib/Target/LLVMIR/DebugImporter.cpp

```diff
@@ -25,6 +25,269 @@ using namespace mlir;
+    // unbound self-refs to be replaced before being used.
+    for (DIRecursiveTypeAttrInterface selfRef : state.unboundSelfRefs)
+      iter->second.pendingReplacements.try_emplace(selfRef, nullptr);
+
+    return {result, state.unboundSelfRefs.empty()};
+  }
+
+  /// Pop off a frame from the translation stack after a node is done being
+  /// translated.
+  void finally(llvm::DINode *node) {
+    // If `node` is not a potentially recursive type, it will not be on the
+    // translation stack. Nothing to handle in this case.
+    if (translationStack.empty() || translationStack.back().first != node)
+      return;
+
+    // At the end of the stack, all unbound self-refs must be resolved already,
+    // and the entire cache should be accounted for.
+    TranslationState &currLayerState = translationStack.back().second;
+    if (translationStack.size() == 1) {
+      assert(currLayerState.unboundSelfRefs.empty() &&
+             "internal error: unbound recursive self reference at top level.");
+      assert(currLayerState.cacheSize == cache.size() &&
+             "internal error: inconsistent cache size");
+      translationStack.pop_back();
+      cache.clear();
+      return;
+    }
+
+    // Copy unboundSelfRefs down to the previous level.
+    TranslationState &nextLayerState = (++translationStack.rbegin())->second;
+    nextLayerState.unboundSelfRefs.insert(
+        currLayerState.unboundSelfRefs.begin(),
+        currLayerState.unboundSelfRefs.end());
+
+    // The current layer cache is now considered part of the lower layer cache.
+    nextLayerState.cacheSize += currLayerState.cacheSize;
+
+    // Finally pop off this layer when all bookkeeping is done.
+    translationStack.pop_back();
+  }
+
+private:
+  /// Returns the cached result (if exists) with all known replacements applied.
+  /// Also returns the set of unbound self-refs that are unresolved in this
+  /// cached result.
+  /// The cache entry will also be updated with the replaced result and with the
+  /// applied replacements removed from the pendingReplacements map.
+  std::pair<DINodeAttr, llvm::DenseSet<DIRecursiveTypeAttrInterface>>
+  lookup(llvm::DINode *node) {
+    auto cacheIter = cache.find(node);
+    if (cacheIter == cache.end())
+      return {};
+
+    CachedTranslation &entry = cacheIter->second;
+
+    if (entry.pendingReplacements.empty())
+      return std::make_pair(entry.attr,
+                            llvm::DenseSet<DIRecursiveTypeAttrInterface>{});
+
+    mlir::AttrTypeReplacer replacer;
+    replacer.addReplacement(
+        [&entry](DIRecursiveTypeAttrInterface attr)
+            -> std::optional<std::pair<Attribute, WalkResult>> {
+          if (auto replacement = entry.pendingReplacements.lookup(attr)) {
+            // A replacement may contain additional unbound self-refs.
+            return std::make_pair(replacement, mlir::WalkResult::advance());
+          }
+          return std::make_pair(attr, mlir::WalkResult::skip());
```

Comment:
You're absolutely correct... This should definitely recurse. I added a test that would break unless this was recursed on, and fixed it here.
Unfortunately this means any previous runs with this branch was probably producing incorrect IR. So it might be good to check again with this fix (I don't think IR size should differ by too much, since we'll still be reusing IR when this happens). I think I'll also add a followup PR with a verifier for DI types. That way we have some way of verifying correctness.

Reply:
Was just running and I get a strange crash during the replacement. Apparently it tries to get the value of a string attribute somewhere deep in the translation stack (and that attribute seems to be null...). It does not make sense to me at the moment. At the very least I will try to produce a reduced test case later today.

Reply:
I found the issue it is not related to your PR and requires some changes to the Attribute/Type replacer (I will follow up separately). The change did not have a significant impact on performance or file size. So it looks good from that perspective.

Reply:
Can you convert this to a non-recursive algorithm?
Recursions are ticking bombs (https://mlir.llvm.org/getting_started/DeveloperGuide/#style-guide )

Reply:
@joker-eph are you referring to AttrTypeReplacer? I think this sounds like a larger separate project.

Reply:
I'm referring to every single unbounded recursion in the compiler: this is in the dev guide "forever".

Reply:
The debug import follows the same design of the corresponding export. While the usage of recursion is problematic and there should be plans to change it, doing so now is not feasible, I think. This PR resolves a substantial performance problem and we would rather not delay this fix.

Honestly, I'm not sure if anyone currently has capacity to look into turning these recursions into iterative implementations. Non-the-less, we should keep this in mind and try to allocate time to resolve these issues.

Reply:
OK turns out as part of fixing the issue @gysit raised, I had to replace this AttrTypeReplacer usage with a more manual replacer that also tracked recIds. So I made it iterative to avoid this issue (and I think something similar can be done to make AttrTypeReplacer iterative too).

As for the larger issue of the entire DebugImporter & exporter being recursion-based... I agree it's not great but also can be fixed separately. Now that AttrTypeReplacer is not used, this PR doesn't introduce any additional recursion besides what's already there.

Reply:
Thanks!